adapters/webhdfs-adapter/src/main/java/com/connexta/replication/adapters/webhdfs/WebHdfsNodeAdapter.java

```diff
@@ -338,22 +346,53 @@ private String getVersion4Uuid(String fileUrl, Date modificationTime) {
 
   /**
    * Returns the files meeting the criteria for replication by removing elements that: 1) are of
-   * type DIRECTORY or 2) have a modification time before or equal to the filter date, when the
-   * filter date is specified
+   * type DIRECTORY or 2) are not one of the failed IDs and have a modification time before or equal
+   * to the filter date, when the filter date is specified
    *
    * @param files a {@code List} of all {@link FileStatus} objects returned by the GET request
+   * @param failedItemIds a {@code List} of the IDs that failed to be created
    * @param filterDate specifies a point in time such that only files more recent are included; this
    *     value will be set to {@code null} during the first execution of replication
    * @return a resulting {@code List} of {@link FileStatus} objects meeting the criteria
    */
   @VisibleForTesting
-  List<FileStatus> getRelevantFiles(List<FileStatus> files, @Nullable Date filterDate) {
-    files.removeIf(
-        file ->
-            file.isDirectory()
-                || (filterDate != null && !file.getModificationTime().after(filterDate)));
+  List<FileStatus> getRelevantFiles(
+      List<FileStatus> files, List<String> failedItemIds, @Nullable Date filterDate) {
+    List<FileStatus> results = new ArrayList<>();
+    Map<String, FileStatus> oldFiles = new HashMap<>();
+
+    for (FileStatus file : files) {
+      // skip over any directories
+      if (file.isDirectory()) {
+        continue;
+      }
+
+      // files modified after the filter date should always be added to the returned results
+      if (filterDate == null || file.getModificationTime().after(filterDate)) {
+        results.add(file);
+      } else {
+        String fileUrl = getWebHdfsUrl().toString() + file.getPathSuffix();
+        Date modificationTime = file.getModificationTime();
+        String id = getVersion4Uuid(fileUrl, modificationTime);
+
+        oldFiles.put(id, file);
+      }
+    }
+
+    return addFailedItemsToRelevantFiles(results, oldFiles, failedItemIds);
+  }
+
+  private List<FileStatus> addFailedItemsToRelevantFiles(List<FileStatus> relevantFiles, Map<String, FileStatus> oldFiles, List<String> failedItemIds) {
+    // only need to do the lookup if there are items in the failed ID list
+    if (oldFiles.isEmpty() || !failedItemIds.isEmpty()) {
```

Comment:
Is this right? I'm having a hard time with this line for some reason.

Shouldn't we only want to go deeper if oldFiles list _isn't_ empty AND the failedItemIds list isn't empty?

Reply:
It's not right